decred/tests/unit/crypto/secp256k1/test_curve.py

```diff
@@ -3,413 +3,507 @@
+            "03fede65e30b4e7576a2abefc963ddbf9fdccbf791b77c29beadefe49951f7d1",
+            "3",
+            "3f07081927fd3f6dadd4476614c89a09eba7f57c1c6c3b01fa2d64eac1eef31e",
+            "949166e04ebc7fd95a9d77e5dfd88d1492ecffd189792e3944eb2b765e09e031",
+            "eb8cba81bcffa4f44d75427506737e1f045f21e6d6f65543ee0e1d163540c931",
+        ),
+        # Addition with z1!=z2 and z2!=1 same x opposite y.
+        # P(x, y, z) + P(x, -y, z) = infinity
+        (
+            "d3e5183c393c20e4f464acf144ce9ae8266a82b67f553af33eb37e88e7fd2718",
+            "5b8f54deb987ec491fb692d3d48f3eebb9454b034365ad480dda0cf079651190",
+            "2",
+            "dcc3768780c74a0325e2851edad0dc8a566fa61a9e7fc4a34d13dcb509f99bc7",
+            "cafc41904dd5428934f7d075129c8ba46eb622d4fc88d72cd1401452664add18",
+            "3",
+            "0",
+            "0",
+            "0",
+        ),
+        # Addition with z1!=z2 and z2!=1 same point.
+        # P(x, y, z) + P(x, y, z) = 2P
+        (
+            "d3e5183c393c20e4f464acf144ce9ae8266a82b67f553af33eb37e88e7fd2718",
+            "5b8f54deb987ec491fb692d3d48f3eebb9454b034365ad480dda0cf079651190",
+            "2",
+            "dcc3768780c74a0325e2851edad0dc8a566fa61a9e7fc4a34d13dcb509f99bc7",
+            "3503be6fb22abd76cb082f8aed63745b9149dd2b037728d32ebfebac99b51f17",
+            "3",
+            "9f153b13ee7bd915882859635ea9730bf0dc7611b2c7b0e37ee65073c50fabac",
+            "2b53702c466dcf6e984a35671756c506c67c2fcb8adb408c44dd125dc91cb988",
+            "6e3d537ae61fb1247eda4b4f523cfbaee5152c0d0d96b520376833c2e5944a11",
+        ),
+    ]
+
+    for i, (x1, y1, z1, x2, y2, z2, x3, y3, z3) in enumerate(tests):
+        # Convert hex to field values.
+        x1 = curve.FieldVal.fromHex(x1)
+        y1 = curve.FieldVal.fromHex(y1)
+        z1 = curve.FieldVal.fromHex(z1)
+        x2 = curve.FieldVal.fromHex(x2)
+        y2 = curve.FieldVal.fromHex(y2)
+        z2 = curve.FieldVal.fromHex(z2)
+        x3 = curve.FieldVal.fromHex(x3)
+        y3 = curve.FieldVal.fromHex(y3)
+        z3 = curve.FieldVal.fromHex(z3)
+
+        # Ensure the test data is using points that are actually on
+        # the curve (or the point at infinity).
+        assert z1.isZero() or curve.isJacobianOnS256Curve(x1, y1, z1)
+        assert z2.isZero() or curve.isJacobianOnS256Curve(x2, y2, z2)
+        assert z3.isZero() or curve.isJacobianOnS256Curve(x3, y3, z3)
+
+        # Add the two points.
+        fv = curve.FieldVal
+        rx, ry, rz = fv(), fv(), fv()
+        curve.curve.addJacobian(x1, y1, z1, x2, y2, z2, rx, ry, rz)
+        assert rx.equals(x3)
+        assert ry.equals(y3)
+        assert rz.equals(z3)
+
+
+def test_double_jacobian():
+    """
+    TestDoubleJacobian tests doubling of points projected in Jacobian coordinates.
+    """
+    # x1, y1, z1 string // Coordinates (in hex) of point to double
+    #     x3, y3, z3 string // Coordinates (in hex) of expected point
```

Comment:
These comments make more sense next to the variables.

Reply:
Moving them below.